### PR TITLE
TensorRT EP engine cache serialization/deserialization refactor

### PR DESCRIPTION
**Pull request by chilo-ms**
Previous design of TRT EP does engine cache load/save at different time according to whether the sub-graph has dynamic input shape or not. For sub-graph which has dynamic input shape, it handles the cache at inference time, and for sub-graph which has static input shape, it handles cache at EP compile time.
The refactor makes all the engine cache load at EP compile time and save them back at release_state_func.

### CODE REVIEW

**Comment by jywu-msft** [comment]
would be good to run this change through perf dashboard and compare with baseline without these changes.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 1]
```diff
@@ -107,6 +107,7 @@ struct TensorrtFuncState {
   bool engine_decryption_enable;
   int (*engine_decryption)(const char*, char*, size_t*);
   int (*engine_encryption)(const char*, char*, size_t);
+  bool save_engine_to_file; // Whether to save engine to file. Default is false
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h:110]: do we need a separate bool for this?
are there cases when "engine_cache_enable" is set to true but we wouldn't serialize the engines?

**Comment by chilo-ms** [comment]
> would be good to run this change through perf dashboard and compare with baseline without these changes.

That's a good suggestion. Will do.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h (hunk 1 above)
Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h:110]: we might want to rename the variable though.
it's not just serializing engines. there's also profile.
and in future it's possible there may be other metadata we want to write out?

Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: I renamed the variable.
profile cache being updated or not is highly dependent on engine cache. 
I made a comment to that variable for clarity.

Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: any thoughts on where the timing cache serialization would happen when implemented?

Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: My thought is that timing cache serialization is done in release_state_func() same as engine cache serialization.
And we also need another variable update_timing_cache for timing cache serialization/save. 

timing cache serialization/save will happen in following situations:

1. timing cache is enabled, and there is no timing cache present.
2. timing cache is enabled, and timing cache is present, but engine is rebuilt.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**